pkg/reconciler/route/route.go

```diff
@@ -293,16 +293,22 @@ func (c *Reconciler) reconcile(ctx context.Context, r *v1alpha1.Route) error {
 		return err
 	}
 
+	logger.Info("Creating placeholder k8s services")
+	services, err := c.reconcilePlaceholderServices(ctx, r, traffic.Targets)
+	if err != nil {
+		return err
+	}
+
 	logger.Info("Creating ClusterIngress.")
-	desired := resources.MakeClusterIngress(r, traffic, ingressClassForRoute(ctx, r))
+	desired := resources.MakeClusterIngress(ctx, r, traffic, ingressClassForRoute(ctx, r))
 	clusterIngress, err := c.reconcileClusterIngress(ctx, r, desired)
 	if err != nil {
 		return err
 	}
 	r.Status.PropagateClusterIngressStatus(clusterIngress.Status)
 
-	logger.Info("Creating/Updating placeholder k8s services")
-	if err := c.reconcilePlaceholderService(ctx, r, clusterIngress); err != nil {
+	logger.Info("Updating placeholder k8s services with clusterIngress information")
+	if err := c.reconcileServices(ctx, r, services, clusterIngress); err != nil {
```

Comment:
The function `reconcileServices` and `reconcilePlaceholderServices` confuses me a lot. Could we merge these two functions to a single one (probably still name it as`reconcilePlaceholderServices`)? In the single `reconcilePlaceholderServices` function, we do creating/updating/deleting placeholder services to move the placeholder services to the desired state. Then we can still keep the current [pattern](https://github.com/knative/serving/blob/6c6bbaae7790125425e5974313067014f547558e/pkg/reconciler/route/route.go#L319) in `route.go` file.

Reply:
Mm, I think the functionality is different between the two function, but perhaps naming can be updated.

`reconcilePlaceholderServices` deals with creating/deleting the k8s services.
- If the service doesn't exist, create it
- If the service does exist, do I own it?
- If the service isn't one I want, delete it
We don't want to update it back to the placeholder spec because the service will flip flop between placeholder and actual spec causing reconcile to trigger again for the route. 

`reconcileServices` updates any services used in `reconcilePlaceholderServices`
- Update service with information provided by the cluster ingress

Reply:
Correct me if i am wrong. For a service that does not exist, it is firstly created in `reconcilePlaceholderServices` with the shape made with `MakeK8sPlaceholderService` , and then it is updated in the `reconcileServices` with the  shape made with `MakeK8sService`. Why can't we directly create a placeholder service with `MakeK8sService`? Then it will save a update operation.
Basically my suggestion is:
1. reconcile `ClusterIngress` first.
2. call reconcilePlaceholderServices with the `ClusterIngress` info
*  if the desired service does not exist, create it.
*  if the desired service does exist, do I own it?
* if we own the existing desired service, reconcile it to the desired shape.
* if there is any existing service that we don't want, delete it.
This is basically the current pattern we have, and could also simply the case when creating a new service.

Reply:
We want to make sure the subroutes we'll be creating can actually exist or owned by the route. Since the `ClusterIngress` resource gets picked up by the cluster ingress controller and starts making virtual services etc. We want to fail earlier if that route isn't possible.

This comment will provide a bit more details: https://github.com/knative/serving/issues/3419#issuecomment-483340423

Reply:
I see. thanks for info. Then it makes sense. But we may need to find out new names for `reconcilePlaceholderservices` and `reconcileServices` probably. :)

Reply:
I can definitely agree to picking better names!